packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,119 @@
+ * which is the default.
+ *
+ * @see #deployTime()
+ * @see #deployTimeOrAssigned()
+ * @see #of()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically assigned at deploy time.
+   * This is the default if you don't specify a physical name.
+   */
+  public static deployTime(): PhysicalName {
+    return new DeployTimePhysicalName();
+  }
+
+  /**
+   * A physical name that will be assigned at synthesis time if the resource
+   * is used in a cross-environment manner,
+   * and at deploy time otherwise.
+   * This is the default name implicitly created sub-resources
+   * (like IAM Roles) are created with in the Construct Library.
+   * If you're using a Construct in a cross-environment manner,
+   * you need to use either this name,
+   * or a {@link #of() fixed name}.
+   */
+  public static deployTimeOrAssigned(): PhysicalName {
+    return new DeployTimeOrAssignedPhysicalName();
+  }
+
+  /**
+   * A fixed physical name (one that is known statically, at synthesis time).
+   *
+   * @param name the name to assign
+   */
+  public static of(name: string): PhysicalName {
+    return new FixedPhysicalName(name);
+  }
+
+  protected constructor() {
+  }
+
+  /**
+   * Returns the physical name, as a String.
+   * Can be `undefined`, if a physical name has not been provided,
+   * or an opaque Token, if a generated name should be assigned lazily at synthesis time.
+   */
+  public abstract asString(): string | undefined;
+
+  /** @internal */
+  public abstract _resolveForCrossEnv(resource: IResource): void;
+}
+
+// implementations are private to this module,
+// we only surface them through static factory methods
+
+class DeployTimePhysicalName extends PhysicalName {
+  public asString(): string | undefined {
+    return undefined;
+  }
+
+  /** @internal */
+  public _resolveForCrossEnv(resource: IResource): void {
+    // error out - a deploy-time name cannot be used across environments
+    throw new Error(`Cannot use resource '${resource.node.path}' in a cross-environment fashion, ` +
+      "as it doesn't have a physical name set");
+  }
+}
+
+class FixedPhysicalName extends PhysicalName {
+  private readonly name: string;
+
+  constructor(name: string) {
+    super();
+
+    this.name = name;
+  }
+
+  public asString(): string | undefined {
+    return this.name;
+  }
+
+  /** @internal */
+  public _resolveForCrossEnv(_resource: IResource): void {
+    // nothing to do - the name will always be the same
+  }
+}
+
+/**
+ * A class that allows assigning a physical name if the given
+ * resource is used in a cross-environment manner.
+ * Separated into its own file so that it can be exported,
+ * as it's used in {@link CrossEnvironmentToken},
+ * but still kept private to the @aws-cdk/cdk package.
+ */
+export class DeployTimeOrAssignedPhysicalName extends PhysicalName {
+  private name?: string;
+
+  public asString(): string | undefined {
+    return new Token(() => {
```

Comment:
You should create the token once (in the ctor) and reuse the object. Otherwise, it will return a different Token object every time someone called `asString`:

```ts
class DeployTimeOrAssigned extends PhysicalName {
  constructor() {
    this.value = new Token(() => this.name);
  }
}
```

Reply:
Will do.

Reply:
Done.